### ISSUE TITLE
List-type columns should able to write null in parquet file #513

### ISSUE THREAD

**Issue by tongwaiazuma** [parquet-go/parquet-go#7]
### Descriptions:
We have a program that **read json file**, do some operation, then **write to a new parquet file**. Some problems happened when there are list-type columns.

Originally, this list-type may have a null value in the json source(e.g.`{"list":null})`, we want the parquet file to remain this characteristic(remain null if the json source is null). However, we cannot achieve this using the current [parquet struct tag](https://github.com/segmentio/parquet-go/blob/510764ae9e8007c4974dd8a1dc42f63f4bbcac4e/schema.go#L31-L97C1) this library provided.

```go
type RowType1 struct {
	ListTag []int32 `json:"list_tag" parquet:"list_tag,list"`
}

func main() {
	var rows []RowType1
	strs := []string{
		`{}`,
		`{"list_tag":null}`,
		`{"list_tag":[]}`,
		`{"list_tag":[1,2]}`,
	}
	for _, s := range strs {
		var r RowType1
		json.Unmarshal([]byte(s), &r)
		rows = append(rows, r)
	}
	if err := parquet.WriteFile("file.parquet", rows); err != nil {
		log.Fatalln("error")
	}
}

//The result is printed by `pqrs cat file.parquet`
// {list_tag: []} -> expect {list_tag: null}
// {list_tag: []} -> expect {list_tag: null}
// {list_tag: []}
// {list_tag: [1, 2]}
```

#### I have tried the following methods:
- to make the list-type columns optional

```go
type RowType1 struct {
	ListTag []int32 `json:"list_tag" parquet:"list_tag,list,optional"`
}
```

However, it seems to be all rows will be either null or empty lists, no matter whether the origin json file has value (e.g.`{"list":[1,2]}`) or not.(Wonder if list + optional is not excepted, or it is just a bug)

- make the list-type columns a pointer

```go
type RowType1 struct {
	ListTag *[]int32 `json:"list_tag" parquet:"list_tag,list"`
}
```

// panic: list is an invalid parquet tag: list_tag *[]int32 [list]
However, it seems that the list tag currently only support [Slice type](https://github.com/segmentio/parquet-go/blob/efb014788aab7a7b0c9746c5bbd85042badf88b0/schema.go#L773)

### Expected Result:
- have a way to write parquet with a null value for the list-type column

Is there any possibility to achieve it?

**Comment by stoewer** [comment]
I agree that it should be possible to write a parquet file with a `nil` value for a slice / list. Let's have a look at the different options:

```go
type RowType struct {
    FieldList []string 
}
```
This will create the following parquet schema:
```
message RowType {
    repeated binary FieldList (STRING);
}
```
I'm not 100% sure about this, but my interpretation of this schema is that the repeated column `FieldList` either contains values or not and therefore it is not possible to distinguish between `nil` and `[]string{}`.

The second options is to add a struct tag `parquet:",list"` to `FieldList`:
```go
type RowType struct {
    FieldList []string `parquet:",list"`
}
```
For the above `RowType` parquet-go creates in a parquet file with this schema:
```
message RowType {
    required group FieldList {
        repeated group list {
            required binary element (STRING);
        }
    }
}
```
Since `FieldList` and `element` are both required, it is also not possible to distinguish a between `nil` and `[]string{}`.

The third option is to use the struct tag `parquet:",list,optional"`:
 ```go
type RowType struct {
    FieldList []string `parquet:",list,optional"`
}
```
For this `RowType` parquet-go will create in a parquet file with a schema that looks like this:
```
message RowType {
    optional group FieldList {
        repeated group list {
            required binary element (STRING);
        }
    }
}
```
In this schema `FieldList` is marked as optional and therefore it is possible to distinguish between `nil` (`FieldList` is missing) and `[]string{}` (`FieldList` is present, but `list` does not contain values).

If the assumptions above are correct, the issue should be addressed by fixing parquet-go's behavior when using the struct tag `parquet:",list,optional"`. Using both `list` and `optional` together leads to columns containing only `null` values (as already mentioned by @tongwaiazuma)

**Comment by achille-roussel** [comment]
This was fixed by https://github.com/parquet-go/parquet-go/pull/197
